File: sdks/python/linktor/types/conversation.py
"""
Conversation and Message types
"""

from datetime import datetime
from enum import Enum
from typing import Any, Optional
from pydantic import BaseModel, Field

from linktor.types.common import (
    ChannelType,
    ContentType,
    MessageDirection,
    MessageStatus,
    PaginationParams,
)


class ConversationStatus(str, Enum):
    """Conversation status"""

    OPEN = "open"
    PENDING = "pending"
    RESOLVED = "resolved"
    SNOOZED = "snoozed"


class SenderType(str, Enum):
    """Message sender type"""

    CONTACT = "contact"
    AGENT = "agent"
    BOT = "bot"
    SYSTEM = "system"


class MediaContent(BaseModel):
    """Media content in messages"""

    url: str
    mime_type: str = Field(alias="mimeType")
    filename: Optional[str] = None
    size: Optional[int] = None
    caption: Optional[str] = None
    thumbnail_url: Optional[str] = Field(None, alias="thumbnailUrl")

    class Config:
        populate_by_name = True


class LocationContent(BaseModel):
    """Location content"""

    latitude: float
    longitude: float
    name: Optional[str] = None
    address: Optional[str] = None


class ContactContent(BaseModel):
    """Contact content"""

    name: str
    phones: Optional[list[str]] = None
    emails: Optional[list[str]] = None


class ButtonContent(BaseModel):
    """Button content"""

    type: str  # 'reply', 'url', 'call'
    text: str
    payload: Optional[str] = None
    url: Optional[str] = None
    phone: Optional[str] = None


class ListRow(BaseModel):
    """List row"""

    id: str
    title: str
    description: Optional[str] = None


class ListSection(BaseModel):
    """List section"""

    title: str
    rows: list[ListRow]


class TemplateParameter(BaseModel):
    """Template parameter"""

    type: str  # 'text', 'image', 'document', 'video'
    text: Optional[str] = None
    media: Optional[MediaContent] = None


class TemplateComponent(BaseModel):
    """Template component"""

    type: str  # 'header', 'body', 'button'
    parameters: Optional[list[TemplateParameter]] = None


class TemplateContent(BaseModel):
    """Template content"""

    name: str
    language: str
    components: Optional[list[TemplateComponent]] = None


class MessageContent(BaseModel):
    """Message content"""

    text: Optional[str] = None
    media: Optional[MediaContent] = None
    location: Optional[LocationContent] = None
    contact: Optional[ContactContent] = None
    buttons: Optional[list[ButtonContent]] = None
    list_sections: Optional[list[ListSection]] = Field(None, alias="listSections")
    template: Optional[TemplateContent] = None

    class Config:
        populate_by_name = True


class Message(BaseModel):
    """Message model"""

    id: str
    conversation_id: str = Field(alias="conversationId")
    direction: MessageDirection
    content_type: ContentType = Field(alias="contentType")
    content: MessageContent
    status: MessageStatus
    external_id: Optional[str] = Field(None, alias="externalId")
    sender_id: Optional[str] = Field(None, alias="senderId")
    sender_type: SenderType = Field(alias="senderType")
    delivered_at: Optional[datetime] = Field(None, alias="deliveredAt")
    read_at: Optional[datetime] = Field(None, alias="readAt")
    failed_reason: Optional[str] = Field(None, alias="failedReason")
    metadata: Optional[dict[str, Any]] = None
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")

    class Config:
        populate_by_name = True


class Conversation(BaseModel):
    """Conversation model"""

    id: str
    tenant_id: str = Field(alias="tenantId")
    channel_id: str = Field(alias="channelId")
    channel_type: ChannelType = Field(alias="channelType")
    contact_id: str = Field(alias="contactId")
    status: ConversationStatus
    assigned_to: Optional[str] = Field(None, alias="assignedTo")
    assigned_at: Optional[datetime] = Field(None, alias="assignedAt")
    last_message: Optional[Message] = Field(None, alias="lastMessage")
    last_message_at: Optional[datetime] = Field(None, alias="lastMessageAt")
    unread_count: int = Field(alias="unreadCount")
    metadata: Optional[dict[str, Any]] = None
    tags: Optional[list[str]] = None
    created_at: datetime = Field(alias="createdAt")
    updated_at: datetime = Field(alias="updatedAt")

    class Config:
        populate_by_name = True


class ListConversationsParams(PaginationParams):
    """List conversations parameters"""

    status: Optional[ConversationStatus] = None
    channel_id: Optional[str] = Field(None, alias="channelId")
    channel_type: Optional[ChannelType] = Field(None, alias="channelType")
    assigned_to: Optional[str] = Field(None, alias="assignedTo")
    contact_id: Optional[str] = Field(None, alias="contactId")
    tags: Optional[list[str]] = None
    search: Optional[str] = None
    sort_by: Optional[str] = Field(None, alias="sortBy")
    sort_order: Optional[str] = Field(None, alias="sortOrder")

    class Config:
        populate_by_name = True


class SendMessageInput(BaseModel):
    """Send message input"""

    text: Optional[str] = None
    media: Optional[dict[str, Any]] = None
    location: Optional[LocationContent] = None
    buttons: Optional[list[ButtonContent]] = None
    template: Optional[TemplateContent] = None
    metadata: Optional[dict[str, Any]] = None


class UpdateConversationInput(BaseModel):
    """Update conversation input"""

    status: Optional[ConversationStatus] = None
    assigned_to: Optional[str] = Field(None, alias="assignedTo")
    tags: Optional[list[str]] = None
    metadata: Optional[dict[str, Any]] = None

    class Config:
        populate_by_name = True


class ListMessagesParams(PaginationParams):
    """List messages parameters"""

    before: Optional[str] = None
    after: Optional[str] = None


__all__ = [
    "ConversationStatus",
    "SenderType",
    "MediaContent",
    "LocationContent",
    "ContactContent",
    "ButtonContent",
    "ListRow",
    "ListSection",
    "TemplateParameter",
    "TemplateComponent",
    "TemplateContent",
    "MessageContent",
    "Message",
    "Conversation",
    "ListConversationsParams",
    "SendMessageInput",
    "UpdateConversationInput",
    "ListMessagesParams",
]

File: sdks/dotnet/src/Linktor/Types/Conversation.cs
using System.Text.Json.Serialization;

namespace Linktor.Types;

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum ConversationStatus
{
    Open,
    Pending,
    Resolved,
    Closed
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum ConversationPriority
{
    Low,
    Medium,
    High,
    Urgent
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum MessageType
{
    Text,
    Image,
    Video,
    Audio,
    Document,
    Location,
    Contact,
    Sticker,
    Template,
    Interactive,
    System
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum MessageStatus
{
    Pending,
    Sent,
    Delivered,
    Read,
    Failed
}

[JsonConverter(typeof(JsonStringEnumConverter))]
public enum MessageDirection
{
    Inbound,
    Outbound
}

public class Conversation
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("tenantId")]
    public string TenantId { get; set; } = string.Empty;

    [JsonPropertyName("channelId")]
    public string ChannelId { get; set; } = string.Empty;

    [JsonPropertyName("contactId")]
    public string ContactId { get; set; } = string.Empty;

    [JsonPropertyName("assignedAgentId")]
    public string? AssignedAgentId { get; set; }

    [JsonPropertyName("botId")]
    public string? BotId { get; set; }

    [JsonPropertyName("status")]
    public ConversationStatus Status { get; set; }

    [JsonPropertyName("priority")]
    public ConversationPriority? Priority { get; set; }

    [JsonPropertyName("subject")]
    public string? Subject { get; set; }

    [JsonPropertyName("lastMessage")]
    public Message? LastMessage { get; set; }

    [JsonPropertyName("unreadCount")]
    public int UnreadCount { get; set; }

    [JsonPropertyName("tags")]
    public List<string> Tags { get; set; } = new();

    [JsonPropertyName("metadata")]
    public Dictionary<string, object>? Metadata { get; set; }

    [JsonPropertyName("firstMessageAt")]
    public DateTime? FirstMessageAt { get; set; }

    [JsonPropertyName("lastMessageAt")]
    public DateTime? LastMessageAt { get; set; }

    [JsonPropertyName("resolvedAt")]
    public DateTime? ResolvedAt { get; set; }

    [JsonPropertyName("createdAt")]
    public DateTime CreatedAt { get; set; }

    [JsonPropertyName("updatedAt")]
    public DateTime UpdatedAt { get; set; }
}

public class Message
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("conversationId")]
    public string ConversationId { get; set; } = string.Empty;

    [JsonPropertyName("type")]
    public MessageType Type { get; set; }

    [JsonPropertyName("direction")]
    public MessageDirection Direction { get; set; }

    [JsonPropertyName("status")]
    public MessageStatus Status { get; set; }

    [JsonPropertyName("text")]
    public string? Text { get; set; }

    [JsonPropertyName("media")]
    public MediaContent? Media { get; set; }

    [JsonPropertyName("location")]
    public LocationContent? Location { get; set; }

    [JsonPropertyName("contact")]
    public ContactContent? Contact { get; set; }

    [JsonPropertyName("template")]
    public TemplateContent? Template { get; set; }

    [JsonPropertyName("interactive")]
    public InteractiveContent? Interactive { get; set; }

    [JsonPropertyName("senderId")]
    public string? SenderId { get; set; }

    [JsonPropertyName("senderType")]
    public string? SenderType { get; set; }

    [JsonPropertyName("externalId")]
    public string? ExternalId { get; set; }

    [JsonPropertyName("metadata")]
    public Dictionary<string, object>? Metadata { get; set; }

    [JsonPropertyName("createdAt")]
    public DateTime CreatedAt { get; set; }

    [JsonPropertyName("updatedAt")]
    public DateTime UpdatedAt { get; set; }
}

public class MediaContent
{
    [JsonPropertyName("url")]
    public string Url { get; set; } = string.Empty;

    [JsonPropertyName("mimeType")]
    public string? MimeType { get; set; }

    [JsonPropertyName("filename")]
    public string? Filename { get; set; }

    [JsonPropertyName("size")]
    public long? Size { get; set; }

    [JsonPropertyName("caption")]
    public string? Caption { get; set; }
}

public class LocationContent
{
    [JsonPropertyName("latitude")]
    public double Latitude { get; set; }

    [JsonPropertyName("longitude")]
    public double Longitude { get; set; }

    [JsonPropertyName("name")]
    public string? Name { get; set; }

    [JsonPropertyName("address")]
    public string? Address { get; set; }
}

public class ContactContent
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("phones")]
    public List<PhoneNumber> Phones { get; set; } = new();

    [JsonPropertyName("emails")]
    public List<string> Emails { get; set; } = new();
}

public class PhoneNumber
{
    [JsonPropertyName("type")]
    public string Type { get; set; } = string.Empty;

    [JsonPropertyName("number")]
    public string Number { get; set; } = string.Empty;
}

public class TemplateContent
{
    [JsonPropertyName("name")]
    public string Name { get; set; } = string.Empty;

    [JsonPropertyName("language")]
    public string Language { get; set; } = string.Empty;

    [JsonPropertyName("components")]
    public List<TemplateComponent> Components { get; set; } = new();
}

public class TemplateComponent
{
    [JsonPropertyName("type")]
    public string Type { get; set; } = string.Empty;

    [JsonPropertyName("parameters")]
    public List<TemplateParameter> Parameters { get; set; } = new();
}

public class TemplateParameter
{
    [JsonPropertyName("type")]
    public string Type { get; set; } = string.Empty;

    [JsonPropertyName("text")]
    public string? Text { get; set; }

    [JsonPropertyName("image")]
    public MediaContent? Image { get; set; }

    [JsonPropertyName("document")]
    public MediaContent? Document { get; set; }

    [JsonPropertyName("video")]
    public MediaContent? Video { get; set; }
}

public class InteractiveContent
{
    [JsonPropertyName("type")]
    public string Type { get; set; } = string.Empty;

    [JsonPropertyName("header")]
    public InteractiveHeader? Header { get; set; }

    [JsonPropertyName("body")]
    public InteractiveBody? Body { get; set; }

    [JsonPropertyName("footer")]
    public InteractiveFooter? Footer { get; set; }

    [JsonPropertyName("action")]
    public InteractiveAction? Action { get; set; }
}

public class InteractiveHeader
{
    [JsonPropertyName("type")]
    public string Type { get; set; } = string.Empty;

    [JsonPropertyName("text")]
    public string? Text { get; set; }

    [JsonPropertyName("image")]
    public MediaContent? Image { get; set; }

    [JsonPropertyName("video")]
    public MediaContent? Video { get; set; }

    [JsonPropertyName("document")]
    public MediaContent? Document { get; set; }
}

public class InteractiveBody
{
    [JsonPropertyName("text")]
    public string Text { get; set; } = string.Empty;
}

public class InteractiveFooter
{
    [JsonPropertyName("text")]
    public string Text { get; set; } = string.Empty;
}

public class InteractiveAction
{
    [JsonPropertyName("buttons")]
    public List<Button> Buttons { get; set; } = new();

    [JsonPropertyName("sections")]
    public List<Section> Sections { get; set; } = new();
}

public class Button
{
    [JsonPropertyName("type")]
    public string Type { get; set; } = string.Empty;

    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("title")]
    public string Title { get; set; } = string.Empty;
}

public class Section
{
    [JsonPropertyName("title")]
    public string Title { get; set; } = string.Empty;

    [JsonPropertyName("rows")]
    public List<SectionRow> Rows { get; set; } = new();
}

public class SectionRow
{
    [JsonPropertyName("id")]
    public string Id { get; set; } = string.Empty;

    [JsonPropertyName("title")]
    public string Title { get; set; } = string.Empty;

    [JsonPropertyName("description")]
    public string? Description { get; set; }
}

// Input classes

public class ListConversationsParams : PaginationParams
{
    [JsonPropertyName("status")]
    public ConversationStatus? Status { get; set; }

    [JsonPropertyName("priority")]
    public ConversationPriority? Priority { get; set; }

    [JsonPropertyName("channelId")]
    public string? ChannelId { get; set; }

    [JsonPropertyName("contactId")]
    public string? ContactId { get; set; }

    [JsonPropertyName("assignedAgentId")]
    public string? AssignedAgentId { get; set; }

    [JsonPropertyName("tag")]
    public string? Tag { get; set; }

    [JsonPropertyName("search")]
    public string? Search { get; set; }
}

public class SendMessageInput
{
    [JsonPropertyName("text")]
    public string? Text { get; set; }

    [JsonPropertyName("type")]
    public MessageType? Type { get; set; }

    [JsonPropertyName("media")]
    public MediaContent? Media { get; set; }

    [JsonPropertyName("location")]
    public LocationContent? Location { get; set; }

    [JsonPropertyName("contact")]
    public ContactContent? Contact { get; set; }

    [JsonPropertyName("template")]
    public TemplateContent? Template { get; set; }

    [JsonPropertyName("interactive")]
    public InteractiveContent? Interactive { get; set; }

    [JsonPropertyName("metadata")]
    public Dictionary<string, object>? Metadata { get; set; }

    public static SendMessageInput TextMessage(string text) => new()
    {
        Text = text,
        Type = MessageType.Text
    };
}

public class UpdateConversationInput
{
    [JsonPropertyName("status")]
    public ConversationStatus? Status { get; set; }

    [JsonPropertyName("priority")]
    public ConversationPriority? Priority { get; set; }

    [JsonPropertyName("assignedAgentId")]
    public string? AssignedAgentId { get; set; }

    [JsonPropertyName("tags")]
    public List<string>? Tags { get; set; }

    [JsonPropertyName("metadata")]
    public Dictionary<string, object>? Metadata { get; set; }
}
